File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   val: number
 *   next: ListNode | null
 *   constructor(val?: number, next?: ListNode | null) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 *   }
 * }
 */

function detectCycle(head: ListNode | null): ListNode | null {
  const nodeSet: Set<ListNode> = new Set()
  let curr: ListNode | null = head
  while (curr) {
    if (nodeSet.has(curr)) return curr
    nodeSet.add(curr)
    curr = curr.next
  }
  return null
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode} head
# @return {ListNode}
def detectCycle(head)
  node_set, curr = Set.new, head
  while curr
    return curr if node_set.include?(curr)
    node_set << curr
    curr = curr.next
  end
  nil
end